### PR TITLE
Add support for JSON responses in the webserver

### PR DESCRIPTION
**Pull request by kisabaka**
This PR allows using the webserver to serve JSON results so it can be used as an API backend in a small deployment. In addition to that, it also returns context lines to make it easy to build a simple client around the webserver.

## Issue
https://github.com/google/zoekt/issues/126

## Rationale
The zoekt webserver makes it easy to launch and test a local instance. However, it cannot be used as an API backend for custom web clients since it only returns results in server rendered HTML pages. Additionally, introducing context lines makes it easy to write clients that behave differently from the default webserver.

## Tests
Tested locally with the webserver, and passing `?format=json&ctx=2`. All existing tests pass

### CODE REVIEW

**Comment by keegancsmith** [comment]
@kisabaka thanks will take a look soon. I just pushed a merge of gerrit/master to our master. Can you rebase your changes ontop of our current master so this PR only contains your changes?

**Comment by kisabaka** [comment]
> @kisabaka thanks will take a look soon. I just pushed a merge of gerrit/master to our master. Can you rebase your changes ontop of our current master so this PR only contains your changes?

Done!

@keegancsmith let me know if I should break this up in 2 PRs: Adding context lines and adding JSON API support.

File: eval.go [hunk 1]
```diff
@@ -154,6 +154,9 @@ func (o *SearchOptions) SetDefaults() {
 	if o.TotalMaxImportantMatch == 0 {
 		o.TotalMaxImportantMatch = 10 * o.ShardMaxImportantMatch
 	}
+	if o.NumContextLines > 5 {
+		o.NumContextLines = 5
+	}
```
Comment (keegancsmith) [inline eval.go]: this isn't enforcing a default, but clamping. I also am not sure why you would ignore what the client asks for. What is the motivation? I think if you are worried about abuse, the more appropriate place for this is in the http layer and returning a badrequest response than silently clamping.

File: api.go [hunk 2]
```diff
@@ -80,6 +80,9 @@ type LineMatch struct {
 	LineEnd    int
 	LineNumber int
 
+	LinesBefore [][]byte
+	LinesAfter  [][]byte
```
Comment (keegancsmith) [inline api.go]: FYI you now need to include a copy of the byte slices in `copyFiles` function. The underlying file content is mmapped in, so go's garbage collector doesn't see references to it. As such once we have decided on the results to send to the user we call `copyFiles` to ensure the bytes live beyond the search.

Maybe also document that these fields are only set if `NumContextLines` is set.

File: web/api.go [hunk 3]
```diff
@@ -36,7 +41,7 @@ type ResultInput struct {
 	Stats         zoekt.Stats
 	Duration      time.Duration
 	FileMatches   []*FileMatch
-	SearchOptions string
+	SearchOptions zoekt.SearchOptions
```
Comment (keegancsmith) [inline web/api.go]: I don't see any updates to the templates, either this field was never read (so you don't need it) or there is a runtime bug when rendering the template. If go just magically converts this to a string, I wonder if the output has degraded. IE looks ugly / includes the defaults rather than what the user specified.

File: contentprovider.go [hunk 4]
```diff
@@ -212,6 +212,31 @@ func (p *contentProvider) fillContentMatches(ms []*candidateMatch) []LineMatch {
...
+				}
+				lineNum += 1
+				if (lineNum < num) && (num-lineNum <= numContextLines) {
+					before = append(before, data[prev+1:next])
+				}
+				if (lineNum > num) && (lineNum-num <= numContextLines) {
+					after = append(after, data[prev+1:next])
+				}
+				prev = next
+			}
```
Comment (keegancsmith) [inline contentprovider.go]: this code seems very inefficient since you always start at the beginning of the file. I recommend two possible alternatives:
- have a frontier that goes backwards numContextLines times using `bytes.LastIndexByte` to calculate the before lines. Then the reverse of that for going forward. Start at lineStart and lineEnd respectively
- You can probably use `p.newlines` above (line 168) for faster operations. However, I assume you usually have low numContextLines so the first approach is likely better.

**Review by keegancsmith — CHANGES_REQUESTED** [review]
Thank you for contributing. Can you include some tests that the feature works / we won't regress?

File: api.go (hunk 2 above)
Comment (kisabaka) [reply api.go]: Thanks! I've added them but not sure how to test it properly

File: contentprovider.go (hunk 4 above)
Comment (kisabaka) [reply contentprovider.go]: Thanks, using `p.newlines` turned out to be more readable IMO

File: web/api.go (hunk 3 above)
Comment (kisabaka) [reply web/api.go]: Yes the field was never read in the templates from what I can see but I'm using it in a project I'm experimenting with. I'm fine with removing it though. WDYT?

**Comment by kisabaka** [comment]
> Can you include some tests that the feature works / we won't regress?
I've added some web e2e tests. Please let me know if you want more tests and where will be the proper place to put them.

@keegancsmith PTAL. Thanks for the review

File: web/api.go (hunk 3 above)
Comment (keegancsmith) [reply web/api.go]: Yeah if it wasn't used then lets remove it :)

File: api.go [hunk 5]
```diff
@@ -80,6 +80,10 @@ type LineMatch struct {
 	LineEnd    int
 	LineNumber int
 
+	// LinesBefore and LinesAfter are only set when SearchOptions.NumContextLines is > 0
+	LinesBefore [][]byte
+	LinesAfter  [][]byte
```
Comment (keegancsmith) [inline api.go]: design: is it not easier to just have `Before []byte`? IE reading your code which does this extraction it would be much easier to just create a single `[]byte` rather than per line. Same goes for After

File: contentprovider.go [hunk 6]
```diff
@@ -212,6 +212,47 @@ func (p *contentProvider) fillContentMatches(ms []*candidateMatch) []LineMatch {
...
+				endIndex = p.fileSize
+				if curLine < numLines-1 {
+					endIndex = newLinesIndexes[curLine+1]
+				}
+				after = append(after, data[startIndex+1:endIndex])
+				curLine += 1
+			}
+
+			finalMatch.LinesAfter = after
+		}
```
Comment (keegancsmith) [inline contentprovider.go:218]: This code is quite verbose, I think it would benefit from a helper function maybe. Here is something I just sketched up without testing. Would allow you to share implementation between calculating before and after.

```go
func lines(data []byte, newlines []uint32, low, high int) [][]byte {
    if len(newlines) == 0 {
        return nil
    }

    if low < 0 {
        low = 0
    }
    if high > len(newlines) {
        high = len(newlines)
    }

    var lines [][]byte
    for i := low; i < high; i++ {
        begin := newlines[i] + 1
        if begin > len(data) {
            break
        }

        var end uint32
        if i + 1 < len(newlines) {
            end = newlines[i+1]
        } else {
            end = uint32(len(data))
        }

        lines = append(lines, data[begin:end])
    }

    return lines
}

File: web/e2e_test.go [hunk 7]
```diff
@@ -259,6 +259,369 @@ func checkNeedles(t *testing.T, ts *httptest.Server, req string, needles []strin
...
+							"sixth example",
+							"seventh",
+						},
+					},
+				},
+			},
+		},
+	} {
+		checkResultMatches(t, ts, req, needles)
+	}
```
Comment (keegancsmith) [inline web/e2e_test.go]: Thank you for adding so many tests!

Three more test cases please:

- filename matches with context
- non-zero ctx with a file which ends with a newline.
- non-zero ctx and empty lines in the context

File: web/e2e_test.go [hunk 8]
```diff
@@ -259,6 +259,369 @@ func checkNeedles(t *testing.T, ts *httptest.Server, req string, needles []strin
...
+}
+
+func fragmentsPartiallyEqual(a, b []Fragment) bool {
```
Comment (keegancsmith) [inline web/e2e_test.go]: why can't you just use reflect.DeepEqual on Fragment? Even better, can you use `cmp.Diff` for better display when this fails?

**Review by keegancsmith — COMMENTED** [review]
a suggested simplification in the code inline, but close to merging :) thank you

File: api.go (hunk 5 above)
Comment (kisabaka) [reply api.go]: You're right, it's easier to do that!

File: web/api.go (hunk 3 above)
Comment (kisabaka) [reply web/api.go]: Done!

File: web/e2e_test.go (hunk 8 above)
Comment (kisabaka) [reply web/e2e_test.go]: True! Thanks for the `cmp.Diff` tip. I'm still learning golang :)

File: contentprovider.go (hunk 6 above)
Comment (kisabaka) [reply contentprovider.go:218]: Right! Getting rid of the `[][]byte` also simplifies the code a lot

File: web/e2e_test.go (hunk 7 above)
Comment (kisabaka) [reply web/e2e_test.go]: Added tests. Though one thing you'll notice is that I'm not swallowing the final `\n` in `Before`, `After` and `Fragments` (we already don't return the one in the fragments anyway).

So if I have ctx=1 `\nfoo\n<match>\nbar\n` then the result will be `before=foo` and after=`bar`.

**Review by kisabaka — COMMENTED** [review]
@keegancsmith PTAL. Changed the new fields to be `string` instead of `[]bytes`

File: contentprovider.go [hunk 9]
```diff
@@ -212,6 +212,13 @@ func (p *contentProvider) fillContentMatches(ms []*candidateMatch) []LineMatch {
 		}
 		finalMatch.Line = data[lineStart:lineEnd]
 
+		if numContextLines > 0 {
+			finalMatch.Before = getLines(
+				data, p.newlines(), num-numContextLines-2, num-2)
```
Comment (keegancsmith) [inline contentprovider.go]: why -2? This seems a bit strange? I'd expect either -1 (to translate between 0-based and 1-based indexing). -2 is a bit of a smell in this helper. Also given this is the only helper, we should use the same "based" indexing to avoid translation.

I'd expect the call to be:

```go
getLines(data, p.newlines(), num - numContextLines, num).
```

IE the `low` is inclusive, the `high` is exclusive. Just like slicing in go.

File: contentprovider.go [hunk 10]
```diff
@@ -212,6 +212,13 @@ func (p *contentProvider) fillContentMatches(ms []*candidateMatch) []LineMatch {
...
+			finalMatch.After = getLines(
+				data, p.newlines(), num-1, num+numContextLines-1)
```
Comment (keegancsmith) [inline contentprovider.go]: Similiar to above, this is what I'd expect this to look like:

```go
getLines(data, p.newlines(), num + 1, num + numContextLines)
```

File: contentprovider.go [hunk 11]
```diff
@@ -234,6 +241,24 @@ func (p *contentProvider) fillContentMatches(ms []*candidateMatch) []LineMatch {
 	return result
 }
 
+func getLines(data []byte, newLines []uint32, low, high int) []byte {
```
Comment (keegancsmith) [inline contentprovider.go:244]: can you add some unit tests for this? I worry about all the corner cases :)

File: contentprovider.go [hunk 12]
```diff
@@ -234,6 +241,24 @@ func (p *contentProvider) fillContentMatches(ms []*candidateMatch) []LineMatch {
...
+	if high < 0 || low >= len(newLines) || len(newLines) == 0 {
+		return nil
+	}
+
+	var startIndex uint32
+	if low < 0 {
+		startIndex = 0
+	} else {
+		startIndex = newLines[low] + 1
```
Comment (keegancsmith) [inline contentprovider.go:257]: if `low` is `len(newLines)-1` and the last element of `data` is `\n`, then doing +1 here will do an out of bounds on data.

File: contentprovider.go [hunk 13]
```diff
@@ -234,6 +241,24 @@ func (p *contentProvider) fillContentMatches(ms []*candidateMatch) []LineMatch {
...
+	}
+
+	if high >= len(newLines) {
+		return data[startIndex:]
+	}
+	return data[startIndex:newLines[high]]
```
Comment (keegancsmith) [inline contentprovider.go:263]: nice, this part LGTM. Now before this we just need to make sure we are observing the correct invariants.

File: shards/shards.go [hunk 14]
```diff
@@ -749,6 +749,8 @@ func copyFiles(sr *zoekt.SearchResult) {
 		copySlice(&sr.Files[i].Checksum)
 		for l := range sr.Files[i].LineMatches {
 			copySlice(&sr.Files[i].LineMatches[l].Line)
+			copySlice(&sr.Files[i].LineMatches[l].Before)
+			copySlice(&sr.Files[i].LineMatches[l].After)
```
Comment (keegancsmith) [inline shards/shards.go:756]: minor: can you specialize copySlice to be a noop when the input is nil. Right now you will allocate an empty slice with this call, even though NumContextLines is not set.

**Review by keegancsmith — CHANGES_REQUESTED** [review]
Hey, sorry some more inline feedback. I'll review quickly! In fact stuff is quite minor here, so if you are happy with it I'm happy to follow up by pushing to this branch if I have any more minor feedback so we can get this landed :)

File: contentprovider.go (hunk 9 above)
Comment (kisabaka) [reply contentprovider.go]: So we're looking at intervals (`p.newlines()`) to find the context.
We want `high` to be the index of the line preceding `num`.
`num` is 1-based and `p.newlines()` is 0-based positions of `\n` characters, if we do `num-1` then `high` will be the index of the `\n` where the matched line ends, which is not what we want. We want the index of the `\n` where the matched line begins.

If we have to try to follow go's slicing convention then we'll end up with even more complicated code. The best option IMO is to just add a comment explaining what's happening here. WDYT?

File: contentprovider.go (hunk 10 above)
Comment (kisabaka) [reply contentprovider.go]: Same as above, in the case of `After` we want to start with `low` set after index 0 of `newLines`

File: contentprovider.go (hunk 11 above)
Comment (kisabaka) [reply contentprovider.go:244]: I wrote this case: https://gist.github.com/kisabaka/f017b7305e8057fd223dc8a4ec0419d1

Let me know if you think this isn't a good test :)

File: contentprovider.go (hunk 12 above)
Comment (kisabaka) [reply contentprovider.go:257]: So if we have two newlines together then `newLines` will have something like `[... 15 16 ...]`.
Also `low` is guaranteed to be less than the line we're matching since we're always doing `num -1` when computing `After`. So worse case doing +1 will just find the first `\n` in the matching line.

File: shards/shards.go (hunk 14 above)
Comment (kisabaka) [reply shards/shards.go:756]: ack.

File: contentprovider.go (hunk 9 above)
Comment (keegancsmith) [reply contentprovider.go]: @kisabaka Final suggestion before merging, make the inputs to getLines based on 1-based indexing. IE just do `low -= 2` and `high -= 2` at the top of getLines with the helpful comment you added about why -2. This would make the call site much easier to read and reason correct.

**Comment by keegancsmith** [comment]
I wasn't convinced this didn't have a panic in it, so added a simple test. Now, I did adjust the getLines implementation so it would be 1-based, but we should make the helper robust so this would apply even though we have some invariants on what is possible for `num`.

```diff
diff --git a/contentprovider.go b/contentprovider.go
index 2442e3f..0b28fe6 100644
--- a/contentprovider.go
+++ b/contentprovider.go
@@ -213,15 +213,10 @@ func (p *contentProvider) fillContentMatches(ms []*candidateMatch, numContextLin
 		finalMatch.Line = data[lineStart:lineEnd]
 
 		if numContextLines > 0 {
-			// Why "-2"?
-			// "num" is 1-based, and "p.newlines()" returns an array of "\n" positions in the doc.
-			// If we do "num-1" then high will be the index of the "\n" where the matched line ends
-			// which is not what we want.
-			// We want the index of the "\n" where the matched line begins so we "num-2" to get that.
 			finalMatch.Before = getLines(
-				data, p.newlines(), num-numContextLines-2, num-2)
+				data, p.newlines(), num-numContextLines, num)
 			finalMatch.After = getLines(
-				data, p.newlines(), num-1, num+numContextLines-1)
+				data, p.newlines(), num+1, num+1+numContextLines)
 		}
 
 		for _, m := range lineCands {
@@ -246,7 +241,14 @@ func (p *contentProvider) fillContentMatches(ms []*candidateMatch, numContextLin
 	return result
 }
 
+// getLines returns a slice of data containing the lines [low, high). low is
+// 1-based and inclusive. high is exclusive.
 func getLines(data []byte, newLines []uint32, low, high int) []byte {
+	// newlines[0] is the start of the 2nd line. So adjust low and high to be
+	// based on newLines.
+	low -= 2
+	high -= 2
+
 	if high < 0 || low >= len(newLines) || len(newLines) == 0 {
 		return nil
 	}
diff --git a/contentprovider_test.go b/contentprovider_test.go
new file mode 100644
index 0000000..cfe4928
--- /dev/null
+++ b/contentprovider_test.go
@@ -0,0 +1,51 @@
+package zoekt
+
+import (
+	"bytes"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+func TestGetLines(t *testing.T) {
+	data := []byte(`one
+two
+three
+four`)
+
+	var newLines []uint32
+	for i, c := range data {
+		if c == '\n' {
+			newLines = append(newLines, uint32(i))
+		}
+	}
+
+	lines := bytes.Split(data, []byte{'\n'}) // TODO does split group consecutive sep?
+	wantGetLines := func(low, high int) []byte {
+		low--
+		high--
+		if low < 0 {
+			low = 0
+		}
+		if low >= len(lines) {
+			return nil
+		}
+		if high <= 0 {
+			return nil
+		}
+		if high > len(lines) {
+			high = len(lines)
+		}
+		return bytes.Join(lines[low:high], []byte{'\n'})
+	}
+
+	for low := -1; low <= len(lines)+2; low++ {
+		for high := low; high <= len(lines)+2; high++ {
+			want := wantGetLines(low, high)
+			got := getLines(data, newLines, low, high)
+			if d := cmp.Diff(string(want), string(got)); d != "" {
+				t.Fatal(d)
+			}
+		}
+	}
+}
```

File: contentprovider.go (hunk 9 above)
Comment (kisabaka) [reply contentprovider.go]: Sounds good!

**Comment by kisabaka** [comment]
@keegancsmith thanks for the `contentprovider_test.go`. I've added it

File: shards/shards.go [hunk 15]
```diff
@@ -738,6 +738,9 @@ func observeMetrics(sr *zoekt.SearchResult) {
 }
 
 func copySlice(src *[]byte) {
+	if src == nil {
```
Comment (keegancsmith) [inline shards/shards.go]:
```suggestion
	if *src == nil {
```

**Review by keegancsmith — APPROVED** [review]
one very minor bug inline, but let me land this now and I'll follow up with a commit to fix it rather than waiting for a full review cycle :)
